pkg/apis/batch/types.go

```diff
@@ -256,6 +256,47 @@ type PodFailurePolicy struct {
 	Rules []PodFailurePolicyRule
 }
 
+// SuccessPolicy describes when a Job can be declared as succeeded based on the success of some indexes.
+type SuccessPolicy struct {
+	// Criteria represents the list of alternative criteria for the declaring the Jobs
+	// as successful before its completion. Once any of the criteria are met,
+	// the "SucceededCriteriaMet" condition is added, and the lingering pods are removed.
+	// The terminal state for such a Job has the "Completed" condition.
+	// Additionally, these criteria are evaluated in order; ONce the Job meets one of the criteria,
+	// other criteria are ignored.
+	// +listType=atomic
+	Criteria []SuccessPolicyCriteria
+}
+
+// SuccessPolicyCriteria describes criterion for declaring a Job as succeeded.
+// Each criterion must have at least one of the "succeededIndexes" or "succeededCount" specified.
+type SuccessPolicyCriteria struct {
+	// SucceededIndexes specifies the set of indexes
+	// which need to be contained in the actual set of the succeeded indexes for the Job.
+	// The list of indexes must be within 0 to ".spec.completions-1" and
+	// must not contain duplicates. At least one element is required.
+	// The indexes are represented as decimal integers separated by commands.
+	// The number are listed in represented by the first and last element of the series,
+	// separated by a hyphen.
+	// For example, if the completed indexes are 1, 3, 4, 5 and 7, they are
+	// represented as "1,3-5,7".
+	// When this filed is null, this field doesn't default to any value
+	// and is never evaluated at any time.
+	//
+	// +optional
+	SucceededIndexes *string
+
+	// SucceededCount specifies the minimal required size of the actual set of the succeeded indexes
+	// for the Job. When SucceededCount is used along with SucceededIndexes, the check is
+	// constrained only to the set of indexes specified by SucceededIndexes.
+	// When this field is null, this doesn't default to any value and
+	// is never evaluated at any time.
+	// When specified it needs to be a positive integer.
```

Comment:
Should we just use a uint32 then?

Reply:
Does it mean that it is better to mention a int32 in this comment?

Reply:
Or, it is better to use a non pointer uint32?

Reply:
I just meant that we don't support negative int32 so why allow them in the type?

Reply:
I see, but I'm not sure what is the advantage of using the unit32 here since we used to select `int32` for other places.

If we select the unit32 here, we need to cast type in many places.

Reply:
The main advantage is you have more bits for that integer with unsigned. 

It probably doesn't matter at all.

unsigned int provides support up to 4294967295 while int32 is -2147483648 to 2147483647. Since we don't allow negative integers, we would need validation to make sure that we don't set negative values. If we used an unsigned int, it would be handled for us.

But maybe its overkill so take it or leave it.

Reply:
Actually, even if `uint32` using is ok (discussed also here: https://github.com/kubernetes/kubernetes/pull/123412#discussion_r1505746589), then I think here it would not do any good, because the number of indexes is capped by the `completions` field, which is `max int32`, so there is no point setting `succeededCount` higher.

Reply:
> Actually, even if uint32 using is ok from API perspective (discussed also here: https://github.com/kubernetes/kubernetes/pull/123412#discussion_r1505746589), then here it would not do any good, because the number of indexes is capped by the completions field, which is max int32, so there is no point setting succeededCount higher.

It sounds reasonable. Additionally, I think that it is better to use the `int32`, and keep consistency with `.spec.completions` since the `.spec.successPolicy.criteria[*].succeededCount` is always included in `.spec.completions`.

Reply:
Sounds good to me.